lib/dotetch/registry.ts

```diff
@@ -0,0 +1,320 @@
+//     throw error
+//   }
+// }
+
+/**
+ * getConfigManifest
+ * This should pull the blob from the registry after checking head.
+ * GET /v2/<name>/blobs/<digest>
+ * GET example /v2/53b00bed7a4c6897db23eb0e4cf620e3/blobs/sha256:1aa86408ad62437344cee93c2be884ad802fc63e05795876acec6de0bb21f3cc
+ * @param imageUrl
+ * @param token
+ * @param digest
+ * @returns
+ */
+async function getConfigManifest(
+	imageUrl: string,
+	token: string,
+	digest: string,
+): Promise<ConfigManifestV2> {
+	const options = {
+		method: 'GET',
+		url: `${imageUrl}/blobs/${digest}`,
+		headers: {
+			Authorization: `Bearer ${token}`,
+			Accept: 'application/vnd.docker.image.rootfs.diff.tar.gzip',
+			'Docker-Distribution-API-Version': 'registry/2.0',
+		},
+	};
+	try {
+		const { data } = await axios(options);
+		return await data;
+	} catch (error) {
+		console.error('==> getBlob error', error);
+		throw error;
+	}
+}
+
+/**
+ *
+ * @param imageUrl
+ * @param token
+ * @returns
+ */
+async function getManifest(imageUrl: string, token: string): Promise<Manifest> {
+	const options = {
+		method: 'GET',
+		url: `${imageUrl}/manifests/latest`,
+		withCredentials: true,
+		credentials: 'include',
+		headers: {
+			Accept: 'application/vnd.docker.distribution.manifest.v2+json',
+			Authorization: `Bearer ${token}`,
+			'Docker-Distribution-API-Version': 'registry/2.0',
+		},
+	};
+
+	try {
+		const { data } = await axios(options);
+		return data;
+	} catch (error) {
+		console.error('==> getManifest error', error);
+		throw new Error(`==> NOPE did not get registry manifest. CATCH: ${error}`);
+	}
+}
+
+/**
+ * @param image
+ * @param auth
+ * @returns
+ */
+export async function pullManifestsFromRegistry(
+	image: string,
+	api: string,
+	bearer: string,
+): Promise<ManifestInfosFromRegistry> {
+	const parsedImage = dockerParseImage(image);
+	const registryUrl = getRegistryUrl(parsedImage);
+	const imageUrl = getImageUrl(parsedImage);
+
+	// const authRealmServiceResponse = await getRealmResponse(registryUrl, authHeaders)
+
+	const token = await getToken(
+		parsedImage,
+		registryUrl,
+		api,
+		bearer,
+	);
+
+	const manifest = await getManifest(imageUrl, token);
+	
+	const configDigest = manifest?.config?.digest;
+	const digests = manifest?.layers;
+
+	// We are using this manifest as its V2
+	const configManifestV2 = await getConfigManifest(
+		imageUrl,
+		token,
+		configDigest,
+	);
+	const diffIds = await configManifestV2.rootfs.diff_ids;
+
+	const imageId = configDigest;
+	const imageName = image;
+
+	return {
+		manifest,
+		digests,
+		configDigest,
+		configManifestV2,
+		imageId,
+		imageName,
+		diffIds,
+		imageUrl,
+		token,
+	};
+}
+
+async function getToken(
+	parsedImage: DockerParsedImage,
+	registryUrl: string,
+	api: string,
+	bearer: string,
+	tag?: string,
+) {
+	try {
+		const { repository, namespace } = parsedImage;;
+
+		const { data } = await axios({
+			method: 'GET',
+			url: `${api}/auth/v1/token`,
```

Comment:
This should not be here.

The way to make auth work for the general case is to query the registry, get a 401 back and inspect the response headers. You’ll find everything needed to issue a request to an api for authorization.

Reply:
Yes indeed, this shortcut was only valid when this code was part of image maker

Reply:
Rose had done that in an early iteration, I got it back and adapted it for more flexibility.
It's not fully tested yet.